### PR TITLE
Added sentinel per Target. Added test_next_round , test_same_round_di…

### PR DESCRIPTION
**Pull request by dimriou**

### CODE REVIEW

**Comment by dionyziz** [comment]
Fix build

**Comment by dionyziz** [comment]
Fix commit message

File: backend/breach/strategy.py [hunk 1]
```diff
@@ -88,6 +88,9 @@ def _get_unstarted_samplesets(self):
             started=None
         )
 
+    def _get_sentinel(self):
+        return self._victim.target.sentinel
+
```
Comment (dionyziz) [inline backend/breach/strategy.py]: Why do you need a function for this if it's used only once?

File: backend/breach/tests/base.py [hunk 2]
```diff
@@ -34,6 +34,56 @@ def setUp(self):
...
+                candidatealphabet='0',
+                data='bigbignextround'
+            ),
+            SampleSet.objects.create(
+                round=next_round_round,
+                candidatealphabet='1',
+                data='smallround'
+            )
+        ]
+
```
Comment (dionyziz) [inline backend/breach/tests/base.py]: Why do you have these in `setUp` and not in the particular testcases?

File: backend/breach/tests/base.py [hunk 3]
```diff
@@ -34,6 +34,56 @@ def setUp(self):
             )
         ]
 
+        # Different batch
+        self.dif_batch_victim = Victim.objects.create(
+            target=target,
+            sourceip='192.168.10.141',
+            snifferendpoint='http://localhost/'
+        )
```
Comment (dionyziz) [inline backend/breach/tests/base.py]: Why do you need a separate victim?

File: backend/breach/tests/base.py [hunk 4]
```diff
@@ -34,6 +34,56 @@ def setUp(self):
...
+                round=dif_batch_round,
+                candidatealphabet='0',
+                data='exlength'
+            ),
+            SampleSet.objects.create(
+                round=dif_batch_round,
+                candidatealphabet='1',
+                data='length'
+            )
+        ]
```
Comment (dionyziz) [inline backend/breach/tests/base.py]: Not sure what you're testing here. Isn't the test supposed to check that the Strategy model is creating the appropriate samplesets during different phases?

File: backend/breach/tests/test_strategy.py [hunk 5]
```diff
@@ -33,11 +34,72 @@ def test_first_round(self, Sniffer):
...
+        work1 = strategy1.get_work()
+        self.assertTrue(work1)
+        self.assertFalse(strategy1.work_completed())
+
+        samplesets_to_check = SampleSet.objects.filter(
+            round=strategy1._round
+        )
+        
+        # Test if the sampleset's new batch is the same with the old one
+        self.assertEqual(samplesets_to_check[0].round, samplesets_to_check[2].round)
```
Comment (dionyziz) [inline backend/breach/tests/test_strategy.py]: You are asserting these are equal even though you created them in your tests? This isn't testing the implementation.

**Comment by dionyziz** [comment]
Split PR about sentinel and PR about backend tests.

File: backend/breach/tests/base.py [hunk 6]
```diff
@@ -34,6 +34,56 @@ def setUp(self):
...
+                round=next_round_round,
+                candidatealphabet='0',
+                data='bigbignextround'
+            ),
+            SampleSet.objects.create(
+                round=next_round_round,
+                candidatealphabet='1',
+                data='smallround'
+            )
+        ]
```
Comment (dionyziz) [inline backend/breach/tests/base.py]: Similarly about this – you're creating the samplesets that are supposed to be created by the Strategy module. What are you testing exactly?

File: backend/breach/tests/test_strategy.py [hunk 7]
```diff
@@ -33,11 +34,72 @@ def test_first_round(self, Sniffer):
...
+        work1 = strategy1.get_work()
+        self.assertTrue(work1)
+        self.assertFalse(strategy1.work_completed())
+
+        # Check if new samplesets have been created for the next round
+        next_round_samplesets = list()
+        samplesets_to_check = SampleSet.objects.filter()
+
+        for sample in samplesets_to_check:
+            if (sample.round.index==2):
```
Comment (dionyziz) [inline backend/breach/tests/test_strategy.py]: style

File: backend/breach/tests/test_strategy.py [hunk 8]
```diff
@@ -33,11 +34,72 @@ def test_first_round(self, Sniffer):
...
+            if (sample.round.index==2):           
+                next_round_samplesets.append(sample)
+    
+        alphabet = '0123456789'
+        index = 0
+
+        for sampleset in next_round_samplesets:
+            self.assertEqual(sampleset.round.knownsecret, 'testsecret1')
+            self.assertEqual(sampleset.candidatealphabet, alphabet[index])
+            index +=1
```
Comment (dionyziz) [inline backend/breach/tests/test_strategy.py]: style

**Comment by dionyziz** [comment]
Use descriptive PR message

File: backend/breach/tests/test_strategy.py [hunk 9]
```diff
@@ -33,11 +34,72 @@ def test_first_round(self, Sniffer):
...
+
+
+        instance.read.return_value = capture1
+        strategy1 = Strategy(self.next_round_victim)
+        work1 = strategy1.get_work()
+        self.assertTrue(work1)
+        self.assertFalse(strategy1.work_completed())
+
+        # Check if new samplesets have been created for the next round
+        next_round_samplesets = list()
```
Comment (dionyziz) [inline backend/breach/tests/test_strategy.py]: []

File: backend/breach/tests/test_strategy.py [hunk 10]
```diff
@@ -33,11 +34,72 @@ def test_first_round(self, Sniffer):
...
+        )
+        
+        # Test if the sampleset's new batch is the same with the old one
+        self.assertEqual(samplesets_to_check[0].round, samplesets_to_check[2].round)  
+        self.assertEqual(samplesets_to_check[0].candidatealphabet, samplesets_to_check[2].candidatealphabet)  
+        self.assertEqual(samplesets_to_check[0].alignmentalphabet, samplesets_to_check[2].alignmentalphabet)  
+        self.assertNotEqual(samplesets_to_check[0].batch, samplesets_to_check[2].batch)  
+
+    @patch('breach.strategy.Sniffer')
+    def test_advance_round(self, Sniffer):
```
Comment (dionyziz) [inline backend/breach/tests/test_strategy.py]: You are not asserting that the round has been advanced anywhere.

File: backend/breach/tests/base.py (hunk 3 above)
Comment (dimriou) [inline backend/breach/tests/base.py]: Test_decide requires a unique victim_id

File: backend/breach/tests/base.py (hunk 2 above)
Comment (dimriou) [inline backend/breach/tests/base.py]: I create these samplesets in order to check if the next round starts with knownsecret: 'testsecret1'

File: backend/breach/tests/base.py (hunk 4 above)
Comment (dimriou) [inline backend/breach/tests/base.py]: 'exlength' and 'length' have a difference of 2 characters. This leads to a confidence of 1 ( current default threshold is 1< ) , so work_completed creates another batch of samplesets. Then we compare previous samplesets with the new ones

File: backend/breach/tests/base.py (hunk 6 above)
Comment (dimriou) [inline backend/breach/tests/base.py]: I wanted to create an initial state before testing the next round, in order to check whether the new round's samplesets have knownsecret: 'testsecret1'

File: backend/breach/tests/base.py (hunk 2 above)
Comment (dionyziz) [inline backend/breach/tests/base.py]: Move them from `setUp` to the beginning of the respective test case.

File: backend/breach/tests/base.py [hunk 11]
```diff
@@ -34,6 +34,56 @@ def setUp(self):
...
+
+        # Advance round
+        self.next_round_victim = Victim.objects.create(
+            target=target,
+            sourceip='192.168.10.141',
+            snifferendpoint='http://localhost/'
+        )
+        next_round_round = Round.objects.create(
```
Comment (dionyziz) [inline backend/breach/tests/base.py]: Rename var to next_round?

File: backend/breach/tests/test_strategy.py [hunk 12]
```diff
@@ -33,11 +34,72 @@ def test_first_round(self, Sniffer):
...
+
+        instance.read.return_value = capture1
+        strategy1 = Strategy(self.next_round_victim)
+        work1 = strategy1.get_work()
+        self.assertTrue(work1)
+        self.assertFalse(strategy1.work_completed())
+
+        # Check if new samplesets have been created for the next round
+        next_round_samplesets = list()
+        samplesets_to_check = SampleSet.objects.filter()
```
Comment (dionyziz) [inline backend/breach/tests/test_strategy.py]: Do you really need to query the samplesets from the db in this case? Why not use Strategy get_work?

File: backend/breach/tests/test_strategy.py [hunk 13]
```diff
@@ -33,11 +34,72 @@ def test_first_round(self, Sniffer):
...
+
+        for sample in samplesets_to_check:
```
Comment (dionyziz) [inline backend/breach/tests/test_strategy.py]: This filtering would be unnecessary if you used Strategy's get_work.

**Pull request closed without merging**